### DISCUSSION TITLE
Node values are changing shortly after setting

### DISCUSSION THREAD

**Discussion by mstfldmr** [FreeOpcUa/python-opcua#1404] [General]
I use python-opcua to set some values in a Kepware server. I set 11 nodes one after another. I observe that values are initially set correctly, but after a few operations they change.

For example, I set all nodes to 601, then some them (not all) change to 2.004. If I set to another value, the resulting value is different as well.

Can you spot anything wrong in the following code?

What may be changing the values?

```
node_ids = ['ns=2;s=FACTORY.MDF1_ONERI_CEKME_DIRENCI',
'ns=2;s=FACTORY.MDF1_ONERI_EGILME_DIRENCI',
'ns=2;s=FACTORY.MDF1_ONERI_ELASTIKIYET_MODULU',
'ns=2;s=FACTORY.MDF1_ONERI_KENAR_VIDALAMA_GERI_CEKME',
'ns=2;s=FACTORY.MDF1_ONERI_RUTUBET',
'ns=2;s=FACTORY.MDF1_ONERI_SISME_24H',
'ns=2;s=FACTORY.MDF1_ONERI_SU_ALMA_24H',
'ns=2;s=FACTORY.MDF1_ONERI_TUTKAL_ONERILEN_DEGER',
'ns=2;s=FACTORY.MDF1_ONERI_YOGUNLUK',
'ns=2;s=FACTORY.MDF1_ONERI_YUZEY_EMMESI',
'ns=2;s=FACTORY.MDF1_ONERI_YUZEY_SAGLAMLIGI']

def set_node_value(node_id, value):
  dv = ua.DataValue(ua.Variant(value, ua.VariantType.Float))
  dv.ServerTimestamp = None
  dv.SourceTimestamp = None
  my_node.set_value(dv)


new_value = 601.0
for node_id in node_ids:
  set_node_value(node_id, new_value)
  sleep(0.5)
```

**Comment by AndreasHeine** [comment]
does the server overwrite them?

**Reply by mstfldmr** [reply]
yes, but only after setting another node, which is weird.

Can you spot anything wrong in my code?

**Reply by swamper123** [reply]
What I would try is:
- modify the script to write just one node
- let wireshark run
- execute the script
- stop wireshark and check if:
    - the value 601.0 was send
    - the connection was closed
    - the value and behaviour on server side (is the single Node changing?)

If it works with a single node without any problems, try it again with multiple Nodes (like 3-5).

If there aren't any odd things on client side, it has to be a server thing.

EDIT: And if this is a Kepware Server thing, you have to ask Kepware.

**Comment by mstfldmr** [comment]
I simplified the code and investigated a bit deeper. Now, I can summarize it as:

```
      my_node = client.get_node(nodeA)

      dv = ua.DataValue(ua.Variant(601.0, ua.VariantType.Float))
      dv.ServerTimestamp = None
      dv.SourceTimestamp = None

      my_node.set_value(dv)

```

When I run this code, the value of NodeA is set to 601.0, but value of NodeB changes to 2.004 as well. Similarly, when I set value of NodeB to 601.0, value of NodeC changes to 2.004.

When I set value of NodeA to 601.2, NodeB changes to 107618480. Likewise, when I set value of NodeB to 601.2, NodeC changes to 107618480.

When I don't set any values, they remain unchanged.

It feels like some kind of data overflow.

**Reply by zerox1212** [reply]
I don't remember how the binary protocol treats a Float variant, but it's possible it's 64bit.

Why not try writing a data type that is for sure 32 bit and see what happens?

FYI Kepware is generally regarded as terrible UA server.

**Reply by swamper123** [reply]
Float is 32bit, double is 64bit.

**Reply by mstfldmr** [reply]
I'm already writing Float, so it shouldn't be the problem, right?
